### DISCUSSION TITLE
FR: filter which nodes graphviz will display in 'link map'

### DISCUSSION THREAD

**Discussion by quesada** [zim-desktop-wiki/zim-desktop-wiki#1890] [Ideas]
From UX POW, It could be as simple as 'click to hide' once the graph is generated, but I'm pretty sure this is not easy to implement as zim calls graphviz and it's not iterative (that is, altering the graph is not possible once it's generated, other than calling graphviz again which will take time and loose 'context' like zoom level or position in the graph.)

All these JS/electron tools that use graphs that are fully interactive have an advantage here.

But still, zim could perhaps subset the notes that are passed to graphviz before rendering. Not sure how the UX could be. Search results? A tag?

There's a lot of value in working with graphs and I'm sure this is an aspect of notetaking that is not well developed in any tool, so there's opportunities to create a 'killer feature'

Perhaps there's a way to access the graphviz file outside of zim and filter/be interactive there?

**Comment by introt** [comment]
I opened the PR draft #1955. It's a simple 'right-click and refresh to exclude/include' implementation:

![image](https://user-images.githubusercontent.com/26069787/155584299-f8cc9b19-99a7-4a66-9784-2edce4c79e1a.png)

![image](https://user-images.githubusercontent.com/26069787/155584229-8c38520c-c75e-4966-b4c5-a5662c1eb7b0.png)

Should cover at least #1904. The exclude list is per-dialog, and dies alongside it. I also added a button to save the graph as an attachment, here's the same graph opened in `vimdot`:

![image](https://user-images.githubusercontent.com/26069787/155590038-3905fbf8-ee7f-4670-8302-7e242a1bb5c7.png)

**Comment by introt** [comment]
Moving your post #1900 here:

> Testing the same set of nodes:
xdot (zim): 11 seconds pegging the CPU 100%
obsidian: instant, with animations
>
> If zim wants to be competitive in a world of notetaking tools with graphs, we should reconsider the current method (xdot/graphviz)

I added graph depth preferences and timeit to #1955 for benchmarking graph generation, and was surprised how slow `xdot` is compared to whatever semi-interactive viewer `vimdot` calls - exporting a thousand-line graph takes between 100-200 milliseconds and `vimdot`'s preview feels instant, yet LinkMap takes seconds to render. So - `graphviz` can be fast enough, just need a better viewer - I think you linked one in one of your other LinkMap discussions, I'll take a look at it.

Edit: aforementioned editor is https://github.com/ArsMasiuk/qvge

**Reply by introt** [reply]
I'm getting a lot of crashes with qvge, if anyone else has time to play with it, please report back successes, and bugs upstream.
